src/pudl/io_managers.py

```diff
@@ -382,12 +405,17 @@ class PudlSQLiteIOManager(SQLiteIOManager):
     using the :class:`pudl.metadata.classes.Package` class.
     """
 
+    # TODO(rousik): now that this experimentally supports also writing to parquet
+    # as an alternative storage format, we should probably rename this class
+    # to be less sqlite-centric.
+
```

Comment:
I dug deeper into how we deal with io_managers and I'm a bit perplexed by the apparent complexity here. There are several partial implementations, some of which support reads, some of which only writes, and there's more duplicate code than ideally should be, but I think this is a problem for a little later and shouldn't be necessarily solved as part of this.

After some more tinkering, I think that the idea of composite IO manager that simply hands-off reads/writes to either sqlite or parquet should work well and should be relatively clean too.

The only remaining snag is that dagster support for loading configuration from env variables seems kind of... incomplete... in that there doesn't seem to be support for booleans or for supplying default value for when the env variable is not set. 

I was expecting something along the lines of 1. use built-in default, 2. replace this with env-variable if set, and 3. replace env default with custom value in case it is supplied via dagster ui/job configuration, but either I'm misreading or this is not actually supported?!

Reply:
Among us I think @bendnorman knows the most about how Dagster resource configuration is supposed to work. I only have a vague understanding, and am just hoping that we can avoid a proliferation of configuration systems, and ideally use whatever the canonical system is for the framework we've adopted.

Reply:
Do we have to use an env var to configure the `PudlMixedFormatIOManager`? I think just using vanilla boolean attributes would be fine:

```python
write_to_parquet: bool = False
"""If true, data will be written to parquet files."""

read_from_parquet: bool = False
"""If true, data will be read from parquet files instead of sqlite."""
```

We could create a `etl_full_gcp_build` that configures the `PudlMixedFormatIOManager` to write and read from parquet.

You can also set a default value for a `EnvVar` using [`EnvVar.get_value()`](https://github.com/dagster-io/dagster/blob/9aa58e22e68e8455573cc04c7fb6a3448ce933c5/python_modules/dagster/dagster/_config/field_utils.py#L564C16-L568). We could do something like this:

```python
write_to_parquet: bool = bool(EnvVar("PUDL_WRITE_TO_PARQUET").get_value(default=True))
```

Not the most elegant thing ever but I tested it out and it allows you to have a default, read from an env var and overwrite the value in the UI. 

There is an [`IntEnvVar`](https://github.com/dagster-io/dagster/blob/9aa58e22e68e8455573cc04c7fb6a3448ce933c5/python_modules/dagster/dagster/_config/field_utils.py#L502) that casts env var values to ints. Maybe we could make one for booleans?

Reply:
I think my preference is for `PudlMixedFormatIOManager` to be a subclass of `IOManager` and to configure it using the `@io_manager(config_schema=...)` decorator. [These are the legacy concepts](https://docs.dagster.io/concepts/io-management/io-managers-legacy) but I don't think it's great to have our IO managers use a mix of the legacy and current resource concepts.

Eventually we should move to using the new pydantic configurable resources. We explored migrating our setting configurations in https://github.com/catalyst-cooperative/pudl/issues/2842 but decided not to because it was a bigger change than we expected. We could create separate issues for migrating the resources in `pudl.resources` and `pudl.io_managers`.

Reply:
Ah okay, so there *is* a way to do all this configuration using the more ergonomic Pydantic settings system. But we would need to do some migration work to get the system there.

Reply:
@bendnorman `ConfigurableIOManager` is a subclass of `IOManager` and works just fine with the old-style `@io_manager` annotated method. The old-style approach just hands out instance of this io manager to whoever needs it, so I don't think there's a practical problem with how it's done now. Or maybe I'm missing something.

Using `ConfigurableIOManager` will AFAIK make the migration easier once we're ready for it because this is basically already prepared for the new world.

Reply:
This is fair point, the duplication of the config both in the `ConfigurableIOManager` and `config_schema` here would be tedious.

Reply:
@bendnorman, fair point re/ use of `config_schema` here. I will make the necessary changes.

Reply:
Though it's tempting to keep using `ConfigurableIOManager` as that takes care of building constructor for me that would be identical to what I'm about to write :-)